### PR TITLE
Implemented configurable logging through C API #23813.

### PR DESCRIPTION
**Pull request by angelortiz1007**
In CInitOptions added the variable  pub ptrfunc: Option<fn(*const c_char) -> bool> which is responsible for being assigned to an Option<> None, or to a function implemented in the C/C++ code.  The C/C++ implemented function is a platform specific logger.  

In addition, vslogger.rs::log() function was modified to call this C/C++ platform specific function - if implemented - and the Windows C/C++ OutputDebugString() function no longer called.

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #__23813_ (GitHub issue number if applicable)

- [x] There are tests for these changes OR
- [ ] These changes do not require tests because ___

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/24184)

### CODE REVIEW

File: support/hololens/ServoApp/ServoControl/Servo.cpp [hunk 1]
```diff
@@ -69,6 +82,9 @@ Servo::Servo(hstring url, GLsizei width, GLsizei height, float dpi,
   o.enable_subpixel_text_antialiasing = false;
   o.vr_pointer = NULL;
 
+  // Example of how to specify platform specific logger
+  // o.ptrfunc = NULL; //&myptrfunc;
```
Comment (jdm) [inline support/hololens/ServoApp/ServoControl/Servo.cpp]: We want the logger initialized for our application, so no need for any examples - let's create our logging function and initialize it here.

File: ports/libsimpleservo/capi/src/vslogger.rs [hunk 2]
```diff
@@ -31,8 +30,8 @@ impl log::Log for VSLogger {
                 record.target(),
                 record.args()
             );
-            unsafe {
-                OutputDebugStringA(log.as_ptr());
+            if LOG_PTR_FUNC.lock().unwrap().is_some() {
+                LOG_PTR_FUNC.lock().unwrap().unwrap()(log.as_ptr() as *const i8);
```
Comment (jdm) [inline ports/libsimpleservo/capi/src/vslogger.rs]: Let's do this:
```rust
let log_fn = &*LOG_PTR_FUNC.lock().unwrap();
if let Some(log_fn) = *log_fn {
    log_fn(log.as_ptr() as *const c_char);
}
```

File: ports/libsimpleservo/capi/src/lib.rs [hunk 3]
```diff
@@ -276,6 +280,10 @@ unsafe fn init(
         slice::from_raw_parts(opts.vslogger_mod_list, opts.vslogger_mod_size as usize)
     };
 
+    if opts.ptrfunc.is_some() {
```
Comment (jdm) [inline ports/libsimpleservo/capi/src/lib.rs]: No need to check if this is Some or not - we can just set our global regardless.

File: ports/libsimpleservo/capi/src/lib.rs [hunk 4]
```diff
@@ -225,6 +228,7 @@ pub struct CInitOptions {
     pub enable_subpixel_text_antialiasing: bool,
     pub vslogger_mod_list: *const *const c_char,
     pub vslogger_mod_size: u32,
+    pub ptrfunc: Option<fn(*const c_char) -> bool>,
```
Comment (jdm) [inline ports/libsimpleservo/capi/src/lib.rs]: Let's call this `log_func` instead.

**Review by jdm — CHANGES_REQUESTED** [review]

**Comment by jdm** [comment]
@bors-servo r+
Thanks!

**Comment by angelortiz1007** [comment]
You are welcome!

Let me know if you want to have 1 last zoom briefing exit.

**Comment by jdm** [comment]
I won't be available for that, sadly.

**Comment by jdm** [comment]
Superseded by #26196.

**Pull request closed without merging**